---
title: "Recalibrating gene expression fold changes"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Recalibrating gene expression fold changes}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Workflow

```{r setup}
library(recalibrate)
```

The basic recalibration workflow is very simple. A data.frame containing
differential expression per gene (measured as log fold change) is adjusted
using the variance in gene expression metric (V^G).

## Input data

The are two requirements for the input data.frame: The row names have to be
Ensembl gene ids, and one of the columns must contain log fold changes.

As an example, here we are using the first three rows of the Alasoo et al. 2018
dataset of macrophages stimulated with IFNg, which can be downloaded from
[zenodo](https://zenodo.org/records/839011)

```{r}
df <- data.frame(
  log2FoldChange = c(0.304144565986564, 2.0260124809332, 2.50289598633734),
  row.names = c("ENSG00000152213", "ENSG00000154642", "ENSG00000161929")
)
```

## Recalibration

The actual recalibration is then a single function call that returns the
recalibrated fold changes in column `recalibratedFC` of the same data.frame.
Note that the parameter `FC_col_name` needs to be adjusted if nominal fold
changes are not stored in column `log2FoldChange` (DESeq2 default).

```{r}
recalibrateFoldChange(df)
```

# Follow-up analysis example

As an example, here we use the Alasoo et al. 2018 dataset of macrophages
stimulated with IFNg. The file can be downloaded from zenodo, as examplified in
the following via a tempoary file.

```{r}
# Download the file from the URL and save it to the temporary file
library(utils)
url <- "https://zenodo.org/records/839011/files/naive_vs_IFNg_DESeq2_fold_change.txt.gz"
temp_file_path <- tempfile(fileext = ".txt.gz")
download.file(url, destfile = temp_file_path, mode = "wb")
df <- read.table(temp_file_path, header = TRUE, row.names = "gene_id")
# Delete the temporary file
unlink(temp_file_path)
```

Here, we are comparing nominal and recalibrated fold changes per gene:

```{r}
library(ggplot2)

df <- recalibrateFoldChange(df, remove_NA = TRUE)
ggplot(data = df, aes(y = recalibratedFC, x = log2FoldChange)) +
  theme_bw() +
  geom_point(alpha = 0.5, size = 1.2) +
  ylab(bquote("recalibrated fold change (log FC /" ~ sqrt(V^G) ~ ")")) +
  xlab("nominal fold change (log FC)")
```

In our analyses, we generally find some degree of correlation between nominal
and recalibrated fold changes.

```{r}
cor(abs(df[, c("log2FoldChange", "recalibratedFC")]),
  method = "spearman",
  use = "complete.obs"
)
```

## GO term enrichement comparison

The following is an analysis as we performed it in the manuscript that
introduced recalibration. Based on the same dataset as above, it performs
GO enrichment (using the `gprofiler2` package) of genes selected by nominal and
recalibrated fold changes respectively. It then compares enrichment p-values
between the two analysis, colored by wether those terms contain the strings
'response' or 'regulation'.

```{r, fig.height=4, fig.width=7}
top_xgenes <- 2000

# select a background set
bg_genelist <- intersect(rownames(df), rownames(vg_h))

# only select significant genes
df_sig <- subset(df, padj < 0.05)

# select top genes by nominal and recalibrated FC
nfc_order <- order(abs(df_sig$log2FoldChange), decreasing = T)
rfc_order <- order(abs(df_sig$recalibratedFC), decreasing = T)
nfc_genes <- rownames(df_sig)[nfc_order[1:top_xgenes]]
rfc_genes <- rownames(df_sig)[rfc_order[1:top_xgenes]]

# perform GO enrichment
library(gprofiler2)
enrichment <- function(gene.list, bglist) {
  gostres <- gost(
    query = gene.list, organism = "hsapiens", ordered_query = FALSE,
    multi_query = FALSE, significant = TRUE, exclude_iea = FALSE,
    measure_underrepresentation = FALSE, evcodes = FALSE,
    user_threshold = 0.05, correction_method = "g_SCS",
    domain_scope = "custom", custom_bg = bglist,
    numeric_ns = "", sources = "GO", as_short_link = FALSE
  )
  gores <- as.data.frame(gostres$result[, c(3, 11)])
  rownames(gores) <- gores$term_name
  return(gores)
}
gores <- enrichment(nfc_genes, bg_genelist)
rfc_gores <- enrichment(rfc_genes, bg_genelist)

# join enrichments
gores$p_value_rfc <- 1 # impute all non-hits from rfc
for (term in row.names(rfc_gores)) {
  if (!term %in% row.names(gores)) { # add missing GO terms to nfc
    gores[term, "term_name"] <- term
    gores[term, "p_value"] <- 1
  }
  gores[term, "p_value_rfc"] <- rfc_gores[term, "p_value"]
}
rownames(gores) <- gores$term_name

# match GO terms based on strings
termAssociations <- c("regulation", "response")
gores$association <- "other"
for (term in row.names(gores)) {
  for (match in termAssociations) {
    if (grepl(match, term, fixed = T)) {
      if (gores[term, "association"] == "other") {
        gores[term, "association"] <- match
      } else {
        gores[term, "association"] <- paste(gores[term, "association"], match, sep = " & ")
      }
    }
  }
}

# plot GO term associations by method of gene selection
plot.colors <- c(
  "regulation" = "#0000FF", "regulation & response" = "#990099",
  "response" = "#FF0000", "other" = "#999999"
)
ggplot(gores, aes(-log10(p_value), -log10(p_value_rfc), color = association)) +
  geom_point() +
  geom_abline(intercept = 0, slope = 1, linetype = 2) +
  theme_minimal() +
  scale_color_manual(name = "GO Term association", values = plot.colors) +
  scale_x_continuous(name = "selected by nominal FC" ~ -log[10] ~ "p-value") +
  scale_y_continuous(name = "selected by recalibrated FC" ~ -log[10] ~ "p-value") +
  ggtitle(paste("GO enrichment of the top", top_xgenes, "genes")) +
  theme(plot.title = element_text(hjust = 0.5, face = "bold"))
```

We notice that GO terms matching the string regulation are generally more
enriched after recalibration, while GO terms that contain the string 'response'
are more enriched before.

# Acknowledgements

We would like to thank Sarah Kim-Hellmuth, Xiaoting Li, Ryan Collins,
Sanna Gudmundsson, Paul Hoffman, Mariia Minaeva,
and the current and former members of the Lappalainen Lab for helpful
discussions and code sharing.

# Funding

This work was supported by funding from the European Research Council (ERC)
under the European Unionâ€™s Horizon 2020 research and innovation programme
(Grant agreement no. 101043238) and the National Human Genome Research Institute
of the NIH (Grant no. R01GM140287).
Part of the computations were enabled by resources provided by the Swedish
National Infrastructure for Computing (SNIC) at UPPMAX partially funded by the
Swedish Research Council through grant agreement no. 2018-05973.

# Citation

> Rentzsch P, Kollotzek A, Mohammadi P, Lappalainen T (2024)
> Recalibrating differential gene expression by genetic dosage variance prioritizes functionally relevant genes.
> *bioRxiv*,
> [10.1101/2024.04.10.588830](https://doi.org/10.1101/2024.04.10.588830)
